### PR TITLE
Improve CAST('field' AS CHAR(n))

### PR DESCRIPTION
**Pull request by richard67**
Pull Request for Issue #157 (part 1)

### Summary of Changes

Port back https://github.com/joomla/joomla-cms/pull/12417 to the framework.

Thanks @alikon for the CMS PR. I've mentioned you as co-author in the commit of this PR here so some credits go to you.

### Testing Instructions

See testing instructions of https://github.com/joomla/joomla-cms/pull/12417 .

### Documentation Changes Required

No idea.

### CODE REVIEW

**Comment by wilsonge** [comment]
Can you remove the linebreaks between the params please

File: src/Pdo/PdoQuery.php [hunk 1]
```diff
@@ -17,4 +17,32 @@
...
+	 *
+	 * @param   string  $value  The value to cast as a char.
+	 *
+	 * @param   string  $len    The lenght of the char.
+	 *
+	 * @return  string  Returns the cast value.
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function castAsChar($value, $len = null)
```
Comment (wilsonge) [inline src/Pdo/PdoQuery.php:37]: Is this really generic? I thought this would be mysql specific?

Comment (richard67) [reply src/Pdo/PdoQuery.php:37]: Was what had been done in the CMS, I only took it over. Or not?

Comment (richard67) [reply src/Pdo/PdoQuery.php:37]: From what I know, CAST .. AS CHAR is Standard-SQL.

Comment (richard67) [reply src/Pdo/PdoQuery.php:37]: Hmm, it seems we get errors here with tests which we didn't have in the CMS.

**Comment by richard67** [comment]
> Can you remove the linebreaks between the params please

Done.

File: src/Pdo/PdoQuery.php (hunk 1 above)
Comment (richard67) [reply src/Pdo/PdoQuery.php:36]: Error with sqlsrv ... probabbly never tested with that in the CMS.

Comment (richard67) [reply src/Pdo/PdoQuery.php:36]: Sqlsrv problem solved.

Shall we add tests here for that new parameter?

Comment (richard67) [reply src/Pdo/PdoQuery.php:36]: unit tests i mean

Comment (wilsonge) [reply src/Pdo/PdoQuery.php:36]: If you don't mind doing that would be awesome :)

Comment (richard67) [reply src/Pdo/PdoQuery.php:36]: Done.

Marking conversation as resolved.

File: Tests/QueryMysqliTest.php [hunk 2]
```diff
@@ -347,6 +347,12 @@ public function testCastAsChar()
 			$this->equalTo('123'),
 			'The default castAsChar behaviour is quote the input.'
 		);
+
+		$this->assertThat(
+			$q->castAsChar('1234', 3),
+			$this->equalTo('CAST(1234 AS CHAR(3))'),
+			'With optional lenght parameter, castAsChar shall limit the output to that length.'
```
Comment (wilsonge) [inline Tests/QueryMysqliTest.php:354]:
```suggestion
			'With optional length parameter, castAsChar shall limit the output to that length.'
```

File: Tests/QueryPgsqlTest.php [hunk 3]
```diff
@@ -347,6 +347,12 @@ public function testCastAsChar()
 			$this->equalTo('123::text'),
 			'The default castAsChar behaviour is quote the input.'
 		);
+
+		$this->assertThat(
+			$q->castAsChar('1234', 3),
+			$this->equalTo('CAST(1234 AS CHAR(3))'),
+			'With optional lenght parameter, castAsChar shall limit the output to that length.'
```
Comment (wilsonge) [inline Tests/QueryPgsqlTest.php:354]:
```suggestion
			'With optional length parameter, castAsChar shall limit the output to that length.'
```

File: Tests/QueryPostgresqlTest.php [hunk 4]
```diff
@@ -399,6 +399,12 @@ public function testCastAsChar()
 			$this->equalTo('123::text'),
 			'The default castAsChar behaviour is quote the input.'
 		);
+
+		$this->assertThat(
+			$q->castAsChar('1234', 3),
+			$this->equalTo('CAST(1234 AS CHAR(3))'),
+			'With optional lenght parameter, castAsChar shall limit the output to that length.'
```
Comment (wilsonge) [inline Tests/QueryPostgresqlTest.php:406]:
```suggestion
			'With optional length parameter, castAsChar shall limit the output to that length.'
```

File: Tests/QuerySqlsrvTest.php [hunk 5]
```diff
@@ -346,9 +346,15 @@ public function testCastAsChar()
...
 			$q->castAsChar('123'),
-			$this->equalTo('CAST(123 as NVARCHAR(10))'),
+			$this->equalTo('CAST(123 as NVARCHAR(30))'),
 			'The default castAsChar behaviour is quote the input.'
 		);
+
+		$this->assertThat(
+			$q->castAsChar('1234', 3),
+			$this->equalTo('CAST(1234 as NVARCHAR(3))'),
+			'With optional lenght parameter, castAsChar shall limit the output to that length.'
```
Comment (wilsonge) [inline Tests/QuerySqlsrvTest.php:356]:
```suggestion
			'With optional length parameter, castAsChar shall limit the output to that length.'
```

**Review by wilsonge — APPROVED** [review]

File: Tests/QueryMysqliTest.php (hunk 2 above)
Comment (richard67) [reply Tests/QueryMysqliTest.php:354]: How could I not see that.

Comment (richard67) [reply Tests/QueryMysqliTest.php:354]: And at the end of the sentence it was right .. so a kind of multiple choice, chose whichever spelling you like, hehe ;-)

**Comment by richard67** [comment]
Wait, I have to fix more spelling errors.

**Comment by richard67** [comment]
Done.

**Comment by richard67** [comment]
@wilsonge Do you think that these typos which I've corrected in my last commit are worth to be corrected in the CMS repo in staging, too?

**Comment by wilsonge** [comment]
Yes please!

**Comment by richard67** [comment]
I see they have been already corrected there.

**Pull request merged**